### DISCUSSION TITLE
Performance compared to native validation

### DISCUSSION THREAD

**Discussion by knobel-dk** [spatie/laravel-data#875] [Q&A] [answered]
Native Laravel validation performance is well-known to be slow due to `Arr::dot`, see https://github.com/laravel/framework/issues/49375

I was exploring this package hoping that it would be more performant but I get these results:

```php
class IdData extends \Spatie\LaravelData\Data
{
    public function __construct(
        public int $id,
    ) {
    }
}

class ArrayData extends \Spatie\LaravelData\Data
{
    public function __construct(
        /** @var IdData[] */
        public array $array,
    ) {
    }
}

$array = array_map(fn (int $id) => ['id' => $id], range(1, 5_000));

\Illuminate\Support\Benchmark::dd([
    fn () => \Illuminate\Support\Facades\Validator::validate(['array' => $array,], [ 'array.*.id' => 'required|integer', ]),
    fn () => ArrayData::validate(['array' => $array])
]);
```

```bash
array:2 [
  0 => "188.543ms"
  1 => "13,456.497ms"
]
```

Am I doing something wrong?

(And, needless to say, but thank you so much for your contribution to the ecosystem ❤️ )

**Comment by knobel-dk** [comment]
Placing these classes in `App\Data` and running `php artisan data:cache-structures` give a tiny, tiny improvement of a few hundered ms

**Comment by rubenvanassche** [comment]
I think as a rule of thumb, never use data validation for really big collections 😅 

The thing is that we use NestedRules for each data object and those are tailored for every data object since one can always change depending on the values provided (see the rules method). Technically we could add some smart caching here and there which could speed up the process (when there's no rules method we kinda know the rules will probably be the same for every object). 

But adding such a thing is a lot of work and requires really good testing in order to break nothing, something on my long list for data v5 but that will probably take some months to even get started on.

**Reply by knobel-dk** [reply]
Thank you so much for a great reply. Just one question:

> I think as a rule of thumb, never use data validation for really big collections 😅

Any ideas on how to validate a really big collection? 😅

**Reply by rubenvanassche** [reply, marked as answer]
In Flare we've created a validator per item that was surprisingly faster than expected, otherwise manually writing out rules in PHP or searching for another library?

**Reply by cyppe** [reply]
@rubenvanassche Do you have some details/example of how you mean with "a validator per item" so I can use your approach as inspiration. As I also have performance issues with the validator so looking for some alternative approach and as you know this package better than anyone it would be interesting to know how you worked around it.

**Reply by knobel-dk** [reply]
@cyppe 

An update from me as the OP.

I ported this part of my app to Spring Boot with Kotlin and coroutines for multi threading. I saw a 30x performance gain and am happy.

My Laravel codebase is highly optimized and I love Spatie's work and Laravel at large but newing up type-safe object with validation seems to be one of those problems for which there are better tools.

Here's another example of single-threaded PHP/Laravel not being suited for this:

https://github.com/laravel/framework/issues/49375

**Comment by cyppe** [comment]
Im on a mini-mission to try to fix the performance issues. 

I started with improving Arr:dot with the hope it would resolve laravel-data package validation automatically. But it did not do any major benefit, but the Arr:dot improvements for regular validations is still huge so it's not wasted time and will help with all kinds of laravel validate methods. It actually got merged already: https://github.com/laravel/framework/pull/55495

But again - it did not move the needle very much for laravel-data as I hoped.  So I am working on this PR that seems to work good, but any feedback is welcome. 

It's for cases where there is no rules() method in the Data objects, as they can have runtime conditions hard to improve this way.

https://github.com/spatie/laravel-data/pull/1005

Of course extra interested in @rubenvanassche feedback.

**Reply by knobel-dk** [reply]
Impressive performance gains on the Laravel PR.
